subprojects/core-api/src/main/java/org/gradle/api/artifacts/ConfigurationContainer.java

```diff
@@ -124,4 +127,99 @@ public interface ConfigurationContainer extends NamedDomainObjectContainer<Confi
      * @return The configuration.
      */
     Configuration detachedConfiguration(Dependency... dependencies);
+
+    /**
+     * Registers a {@link ResolvableConfiguration} with an immutable role. Resolvable configurations
+     * are meant to resolve dependency graphs and their artifacts.
+     *
+     * @param name The name of the configuration to register.
+     *
+     * @return A provider which creates a new resolvable configuration.
+     *
+     * @throws InvalidUserDataException If a configuration with the given {@code name} already exists in this container.
```

Comment:
This is the exception type we already throw. It comes from `DefaultNamedDomainObjectCollection#assertCanAdd`

Reply:
Right, I just think it's worth extending that with a narrower custom type.  It might be helpful during migrating away from `maybeCreate` and the like to more easily detect something that already exists.  Especially in conjunction with a more central name-validation policy, which we might need if a goal is to keep the names unique in the future across multiple role-based containers.

Reply:
If we do this for configurations, we should do it everywhere. As such it would belong in another PR

Reply:
I don't think there's a need for it with many/any other containers - this is a unique case in that we have things managed by a single container currently that will be split into multiple containers eventually.  A more specific type would still be a `InvalidUserDataException`, and anyone seeking consistency with other containers could still treat it as such.  I just think it's smart to anticipate future name collision problems IRT this configuration work and plan for them now.